File: leetcode/medium/260. Single Number III.py
# problem : https://leetcode.com/problems/single-number-iii/
# time complexity : O(N)

class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        numDict = dict()  # numDict[num] = cnt

        for num in nums:
            cnt = numDict.get(num, 0)
            numDict[num] = cnt + 1

        ans = []
        for k, v in numDict.items():
            if v == 1:
                ans.append(k)
        return ans

File: leetcode/medium/260. Single Number III.cpp
/**
 * problem : https://leetcode.com/problems/single-number-iii/
 * time complexity : O(N)
 * algorithm : bitmask
 */
 
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int sum = 0;
        int N = nums.size();
        for(int i=0;i<N;i++){
            sum ^= nums[i];
        }
        
        int lastSettedBit = (sum & (sum - 1)) ^ sum;
        int ans = 0;
        for(int i=0;i<N;i++){
            if((int)(lastSettedBit & nums[i]) == 0)
                ans ^= nums[i];
        }
        
        return vector<int>{ ans, sum ^ ans };
    }
};
